Fix gain issue with cloned tracks

daschuer opened the pull request:
This PR fixes #10550 Via a couple of single fixes regarding using the right data at the right time. 

This finally allows to play two cloned tracks with changing tempo without that the leader applies future or outdated tempo changes to the follower.

JoergAtGithub commented:
Does this also fix #11788 ?

daschuer commented:
No, @ywwg has a fix in https://github.com/mixxxdj/mixxx/pull/12431

ywwg commented on src/test/enginesynctest.cpp:
```diff
@@ -1937,29 +1964,18 @@ TEST_F(EngineSyncTest, HalfDoubleThenPlay) {
             m_pChannel1->getEngineBuffer()->m_pSyncControl->getBeatDistance(),
             m_pChannel2->getEngineBuffer()->m_pSyncControl->getBeatDistance(),
             kMaxFloatingPointErrorHighPrecision);
-
```
by removing these ProcessBuffers the two checks at the bottom are identical -- does it fail if play advances?

daschuer replied:
Ah Ok, I did not realize that this was a test. I will revert it.

daschuer commented:
Done.

ronso0 commented:
I confirm this fixes #10550 

But it seems there is a difference in how clones are aligned (without Sync) in 2.4 compared to 2.3.6, there seems to be some tiny offset / flanger.
This zip contains some Flac files for comparison: -solo, -clone_2.3.6 and -clone_2.4-beta
[zeitgeist-clone.tar.gz](https://github.com/mixxxdj/mixxx/files/13728838/zeitgeist-clone.tar.gz)

daschuer commented:
Let's focus on https://github.com/mixxxdj/mixxx/issues/10550 which aims to fix all flanger effects, independent of sync and quantize. 

There is no need for a recording. If you are able to create a flanger, please describe the steps and you may also share the file. Than I can hunt down the issue.

daschuer commented:
There was a similar ssue, that is fixed in https://github.com/mixxxdj/mixxx/issues/10550 does that fix the issue for you?

ronso0 commented:
We're in the PR that fixes #10550 ;)

I noticed the effect with this PR, then I tested with 2.4 (identical) and with 2.3 from the ppa (issue not present).

daschuer commented:
Ah ok. Is it a fixed bpm track?

ronso0 commented:
Nope, variable BPM. Approx. 105-145
I'll write that report and send you the file via email.

daschuer committed:
Improve casting in EnginePregain

daschuer committed:
Don't apply ramping gain when the track is stopped or starting up.

daschuer committed:
Set up replay gain before starting playing

daschuer commented:
I have discovered and fixed a ramping issue with ReplayGain that is now fixed. 

I can reproduce the Flanger issue with Soundtouch + Keylock enabled. My guess is that the fresh initialized filters behave different compared to the established filters of the playing deck. Rubberband does not show the issue. Not sure if we can fix it from the Mixxx code. This is out of scope of this PR. 

@ronso0 can you confirm this?

daschuer commented:
This should also fix the flipping test HotcueControlTest.CueGotoAndPlay on Launchpad, because the result original depends on thread timing.

daschuer commented:
@ronso0 can we merge this now?

ronso0 commented:
I don't understand the EnginePregain changes, sorry.
Diving into that will take time.

What about moving those commits to another PR?
Then we can merge the Sync fix right away.

daschuer commented:
You can find the first part of this PR here: https://github.com/mixxxdj/mixxx/pull/12515

ronso0 commented:
Okay, #12515 targets main, so please remove the pregain commits here.

daschuer commented:
https://github.com/mixxxdj/mixxx/pull/12515 is now for 2.4 with removed gain commits. 
This one can be merged in a second step.

daschuer commented:
@ronso0 Will you find time to repeat you original tests with this branch soon? If not, let's remove this form the 2.4 milestone, since the gain issue is not a 2.4 regression. The sync part also contains in https://github.com/mixxxdj/mixxx/pull/12515 is the 2.4 part.

ywwg commented:
@daschuer now that #12515 is merged, can you update this PR? it's still showing all the diffs instead of the ones unique to this branch

daschuer commented:
Done.

ywwg commented on src/engine/enginepregain.cpp:
```diff
@@ -88,17 +88,17 @@ void EnginePregain::process(CSAMPLE* pInOut, const int iBufferSize) {
             float seconds = static_cast<float>(m_timer.elapsed().toDoubleSeconds());
-            if (seconds < kFadeSeconds) {
-                // Fade smoothly
+            if (seconds < kFadeSeconds && m_dSpeed != 0.0 && m_dOldSpeed != 0.0) {
```
does this mean that once the replaygain is calculated, mixxx will now change the gain while the track is playing back?

daschuer replied:
These condition applies the replay gain instantly without fading if the track is or was stopped. If it plays, fading is done. 
The logic that a analyzer replay gain is not applied when playing is handled here: https://github.com/mixxxdj/mixxx/blob/59ca3a25b768c44ac0934ed4cb4504cac52bde54/src/mixer/basetrackplayer.cpp#L711

ywwg commented on src/test/reference_buffers/CueGotoAndStopTest:
```diff
@@ -1,512 +1,512 @@
-0.00420177,0.00420177
-0.00288188,0.00288188
-0.00147617,0.00147617
+0,0
```
why are these all zeros now?

daschuer replied:
That seems to be wrong. I will have a look.

ywwg commented on src/test/reference_buffers/BasicProcessingTestPause:
```diff
@@ -1,512 +1,512 @@
-0.00245179,0.00245179
-0.00116233,0.00116233
+0,0
+0,0
```
why are these all zeroes now?

ywwg commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -527,7 +527,6 @@ void BaseTrackPlayerImpl::slotTrackLoaded(TrackPointer pNewTrack,
         m_pDuration->set(0);
         m_pFileBPM->set(0);
         m_pKey->set(0);
-        setReplayGain(0);
```
why was this the wrong place to put the replaygain setting?

daschuer replied:
Because it was too late, the first engine call uses the replay gain of the previous loaded track.

daschuer commented:
@ywwg thank you for your good review. The test changes where wrong. 
I have added a fixup commit. If this works for you I will rebase it.

daschuer commented:
@Swiftb0y @ronso0 I think this new fixup! checker suites good to prevent merging preliminary PRs. 
What do you think?

daschuer commented on src/test/reference_buffers/ScratchTestStart:
```diff
@@ -1,6 +1,6 @@
 -0.000926289,-0.000926289
 -0.00185258,-0.00185258
--0.00277897,-0.00277897
+-0.00277896,-0.00277896
```
I will revert this rounding artefact during fixup

Swiftb0y commented:
> @Swiftb0y @ronso0 I think this new fixup! checker suites good to prevent merging preliminary PRs.

Yup, I agree. Is there already PR for that I missed?

daschuer commented:
No, I will create one.

ywwg reviewed (approved):
thank you!

daschuer commented:
All green after debase and an approval from @ywwg. So it is ready to go.
